src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -390,31 +395,79 @@ public static void Register(Type t, DeepCopier cop, Serializer ser, Deserializer
+                                    clone[t.TypeHandle] = cop;
+                                    //Put in a memory barrier to ensure the dictionary is fully updated before we switch it to the new location
+                                    System.Threading.Thread.MemoryBarrier();
+                                    copiers = clone;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            lock (copiers)
+                            {
+                                DeepCopier current;
+                                if (forceOverride || !copiers.TryGetValue(t.TypeHandle, out current) || (current == null))
+                                {
+                                    copiers[t.TypeHandle] = cop;
+                                }
                             }
                         }
                     }
                     if (ser != null)
                     {
-                        lock (serializers)
+                        if (_setupDone)
                         {
-                            Serializer currentSer;
-                            if (forceOverride || !serializers.TryGetValue(t.TypeHandle, out currentSer) || (currentSer == null))
+                            lock (ser)
                             {
-                                serializers[t.TypeHandle] = ser;
+                                Serializer currentSer;
+                                if (forceOverride || !serializers.TryGetValue(t.TypeHandle, out currentSer) || (currentSer == null))
+                                {
+                                    var clone = serializers.ToDictionary(kv => kv.Key, kv => kv.Value);
+                                    clone[t.TypeHandle] = ser;
+                                    //Put in a memory barrier to ensure the dictionary is fully updated before we switch it to the new location
+                                    System.Threading.Thread.MemoryBarrier();
+                                    serializers = clone;
+                                }
+                            }
+                            lock (deserializers)
+                            {
+                                Deserializer currentDeser;
+                                if (forceOverride || !deserializers.TryGetValue(t.TypeHandle, out currentDeser) || (currentDeser == null))
+                                {
+                                    var clone = deserializers.ToDictionary(kv => kv.Key, kv => kv.Value);
+                                    clone[t.TypeHandle] = deser;
+                                    //Put in a memory barrier to ensure the dictionary is fully updated before we switch it to the new location
+                                    System.Threading.Thread.MemoryBarrier();
+                                    deserializers = clone;
+                                }
```

Comment:
just create a generic helper method to add an item to a dictionary, and conditionally do a set of a clone and swap. Otherwise you are duplicating all of the logic for each of the collections